src/libsyntax/diagnostic.rs

```diff
@@ -474,42 +474,56 @@ fn highlight_lines(err: &mut EmitterWriter,
         // how many digits must be indent past?
         while num > 0u { num /= 10u; digits += 1u; }
 
-        // indent past |name:## | and the 0-offset column location
-        let left = fm.name.len() + digits + lo.col.to_uint() + 3u;
         let mut s = String::new();
         // Skip is the number of characters we need to skip because they are
         // part of the 'filename:line ' part of the previous line.
-        let skip = fm.name.len() + digits + 3u;
+        let skip = fm.name.width(false) + digits + 3u;
         for _ in range(0, skip) {
             s.push(' ');
         }
         if let Some(orig) = fm.get_line(lines.lines[0]) {
-            for pos in range(0u, left - skip) {
-                let cur_char = orig.as_bytes()[pos] as char;
+            let mut lastc = ' ';
+            let mut iter = orig.chars().enumerate();
+            for (pos, ch) in iter {
+                lastc = ch;
+                if pos >= lo.col.to_uint() { break; }
                 // Whenever a tab occurs on the previous line, we insert one on
                 // the error-point-squiggly-line as well (instead of a space).
                 // That way the squiggly line will usually appear in the correct
                 // position.
-                match cur_char {
+                match ch {
                     '\t' => s.push('\t'),
-                    _ => s.push(' '),
-                };
+                    c => for _ in 0..c.width(false).unwrap_or(0) {
+                        s.push(' ');
+                    },
+                }
             }
-        }
 
-        try!(write!(&mut err.dst, "{}", s));
-        let mut s = String::from_str("^");
-        let hi = cm.lookup_char_pos(sp.hi);
-        if hi.col != lo.col {
-            // the ^ already takes up one space
-            let num_squigglies = hi.col.to_uint() - lo.col.to_uint() - 1u;
-            for _ in range(0, num_squigglies) {
-                s.push('~');
+            try!(write!(&mut err.dst, "{}", s));
+            let mut s = String::from_str("^");
+            match lastc {
+                // Most terminals render tabs as 8 spaces
+                '\t' => s.push_str("~~~~~~~"),
+                c => for _ in 0..c.width(false).unwrap_or(1)-1 {
```

Comment:
This could be:

```
let count = match ch {
    '\t' => 8,
    _ => ch.width(false).unwrap_or(1) - 1
};
s.extend(std::iter::repeat('~').take(count));
```

And similarly below.